sympy/integrals/integrals.py

```diff
@@ -1242,6 +1242,110 @@ def _sage_(self):
+        >>> Integral(g, (x, -oo, oo)).principal_value()
+        oo
+
+        >>> f = 1 / (x**3)
+
+        >>> Integral(f, (x, -oo, oo)).principal_value()
+        0
+        >>> Integral(f, (x, -10, 10)).principal_value()
+        0
+        >>> Integral(f, (x, -10, oo)).principal_value() + Integral(f, (x, -oo, 10)).principal_value()
+        0
+
+        References
+        ==========
+
+        .. [1] http://en.wikipedia.org/wiki/Cauchy_principal_value
+        .. [2] http://mathworld.wolfram.com/CauchyPrincipalValue.html
+        """
+        from sympy.calculus import singularities
+
+        if len(self.limits) != 1 or len(list(self.limits[0])) != 3:
+            raise ValueError("You need to insert a variable, lower_limit, and upper_limit correctly to calculate "
+                             "cauchy's principal value")
+
+        x, a, b = self.limits[0]
+
+        if not(a.is_comparable and b.is_comparable and a <= b):
+            raise ValueError("The lower_limit must be smaller than or equal to the upper_limit to calculate "
+                             "cauchy's principal value. Also, a and b need to be comparable.")
+
+        if a == b:
+            return 0
+        r = Dummy('r')
+        u = Dummy('u')
+
+        singularities_list = [s for s in singularities(self.function, x) if s.is_comparable and a <= s <= b]
+
+        for i in singularities_list:
+            if ((i == b) or (i == a)):
+                raise ValueError(
+                    'The principal value is not defined in the given interval due to singularity at %d.' % (i))
+
+        f = self.function
+        F = integrate(f, x)
+
+        if F.has(Integral):
+            return self
+        if len(singularities_list) == 0:
+            if a is -oo and b is oo:
+                I = limit(F - F.subs(x, -x), x, oo)
+                return I
+            else:
+                I = limit(F, x, b, '-') - limit(F, x, a, '+')
+                return I
+
+        if len(singularities_list) == 1 and not(a is -oo and b is oo):
+            I = limit(((F.subs(x, singularities_list[0] - r)) - F.subs(x, singularities_list[0] + r)), r, 0,
+                      '+') - limit(F, x, a) + limit(F, x, b)
+            return I
+
+
+        else:
+            if a is -oo and b is oo:
+                Iinf = 0
+                for singular_element in range(len(singularities_list)):
+                    Iinf = Iinf + limit(((F.subs(x, singularities_list[singular_element] - r)) - F.subs(x,
+                                                                                                        singularities_list[
+                                                                                                            singular_element] + r)),
+                                        r, 0, '+') + limit(F.subs(x, u) - F.subs(x, -u), u, oo)
+                return Iinf
+            else:
+                Ib = 0
+                Ia = F.subs(x, singularities_list[0] - r) - limit(F, x, a)
```

Comment:
The first term is added to `Ib` on the first execution of the loop. Later it will be included twice in the sum `Ia + Ib` below.

Reply:
@jksuom  Yes. You are right. I should've seen that. Thanks for this. Also, in the `for` loop, it should iterate in this range : `range(1, len(singularities_list) - 1)` instead of `range(1, len(singularities_list))` ( this I just realised ) since just after the `for` loop, the last element is taken care of.